rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,526 @@
+an API to let workers identify themselves as main or not.
+
+```ts
+export interface ClusteringServiceSetup {
+  // [...]
+  isMainWorker: () => boolean;
+}
+```
+
+Notes:
+- in non-clustered mode, `isMainWorker` would always return true, to reduce the divergence between clustered and 
+  non-clustered modes.
+
+### Example: SO Migration
+
+To take the example of SO migration, the `KibanaMigrator.runMigrations` implementation could change to
+(naive implementation, the function is supposed to return a promise here, did not include that for simplicity):
+
+```ts
+runMigration() {
+   if(clustering.isMainWorker()) {
+     this.runMigrationsInternal().then((result) => {
+        applyMigrationState(result);
+        clustering.broadcast('migration-complete', { payload: result }, { persist: true }); // persist: true will send message even if subscriber subscribe after the message was actually sent
+      })
+   }
+   else {
+     const unsubscribe = clustering.addMessageHandler('migration-complete', ({ payload: result }) => {
+       applyMigrationState(result);
+       unsubscribe();
+     });
+   }
+}
+```
+
+Notes:
+  - to be sure that we do not encounter a race condition with the event subscribing / sending (workers subscribing after 
+    the main worker actually send the `migration-complete` event and then waiting indefinitely), we are using the `persist` option of the `broadcast` API.
+    I'm honestly not sure this is the best pattern, but I couldn't come with anything better (maybe shared state instead?). 
+    This is probably something else we should discuss.
+
+
+## Sharing state between workers
+
+This is not identified as necessary at the moment, and IPC broadcast should be sufficient, hopefully. 
+
+If we do need shared state, we will probably have to use syscall libraries to share buffers such as [mmap-io](https://www.npmjs.com/package/mmap-io), 
+and expose a higher level API for that from the `clustering` service.
+
+# Technical impact on the codebase
+
+This section is meant to be an exhaustive (even if realistically not at the moment) inventory  of the changes that would
+be required to support clustering mode.
+
+## Technical impact on Core
+
+### Handling multi-process logs
+
+This is an example of log output in a 2 workers cluster, coming from the POC:
+
+```
+[2021-03-02T10:23:41.834+01:00][INFO ][plugins-service] Plugin initialization disabled.
+[2021-03-02T10:23:41.840+01:00][INFO ][plugins-service] Plugin initialization disabled.
+[2021-03-02T10:23:41.900+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
+[2021-03-02T10:23:41.903+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
+```
+
+The workers logs are interleaved, and, most importantly, there is no way to see which process each log entry is coming from. 
```

Comment:
Sharing my past experience

As you might be already aware, Without a centralised co-ordinator, the logging to the stdout or any stream would be interleaved this making the output in a broken state if the application becomes chatty or busy. 

One thing that worked very well was to differentiate the master vs the worker mode and use different stream which would eventually write everything to `process.stdout` by subscribing to worker messages. Here is a dummy implementation that worked well

```js
class WorkerStream extends stream.Writable {
    constructor(callback) {
        super({ decodeStrings: false });
        this.callback = callback;
    }

    _write (chunk, _, done) {
        this.callback(chunk);
        done();
    }
}

const workerStream = new WorkerStream((data) => {
    // send to cordinator for logging
    process.send('log', data)
});

function createLogger () {
    const stream = cluster.isWorker ? workerStream : process.stdout
    // abstracted out - could be BaseLogger for Kibana
    return logger(stream)
}

// somewhere in coordinator
process.on('log', data => {
    process.stdout.write(data)
})
```

I don't know much about the Kibana architecture, just sharing my experience working with cluster modules.

Reply:
> As you might be already aware, Without a centralised co-ordinator, the logging to the stdout or any stream would be interleaved this making the output in a broken state if the application becomes chatty or busy

I wonder to which extend this should be considered blocker tbh. Logs from a given process/worker will still be assured to be in order, we'll have the timestamp and the workerId in the entries, and even if flushes between processed are not assured to be in order, the timegap, even under heavy load, is not that significant.

> One thing that worked very well was to differentiate the master vs the worker mode and use different stream which would eventually write everything to process.stdout by subscribing to worker messages

Yea, I listed that in our options. This feels like the proper design. The only problem is that the overall complexity of implementing that is quite high.

> That also addresses the problem with the rolling files. Doesn't it?

Yea, it definitely will. That's option `3.` of the rolling file appender section.

Reply:
I see more than a sorting/ordering problem, an issue when 2+ processes are writing to the same file. They can mixin entries and have broken logs with some content breaking in the middle of some others. 
I think we definitely need the coordinator to ensure log entries are written one at a time.

Reply:
Current proposal is to centralize logging in the coordinator, which should resolve issues of interleaving as the coordinator would be the only process writing to a file or stdout.